_posts/2024-02-23-enhanced-multi-vector-support-in-opensearch-knn.md

```diff
@@ -0,0 +1,182 @@
+With the rise of Large Language Models (LLM), many users are turning to vector databases for indexing, storing, and retrieving information, particularly for building Retrieval-augmented Generation (RAG) systems. These models convert text to embeddings and preserve the semantic information in the embeddings. However, models have limitations on number of tokens to consider for embedding generation which requires document to be chunked and stored as multiple embeddings in a single document. Without multi-vector support, every chunk of the document needs to be treated as a separate entity, requiring duplication of document metadata across multiple chunk unless the user separates the document and its metadata. This duplication of metadata can result in significant storage and memory usage. Furthermore, users must independently devise a merging mechanism to ensure that only a single document is retrieved from multiple chunk during search. Multi-vector support simplifies the handling of large documents, alleviating these challenges.
+
+## Understanding previous limitations
+### Fewer Results
+In previous versions of OpenSearch, when searching for nested KNN (K-Nearest Neighbors) fields with a specified number of nearest neighbors (k value), the search might return fewer than k documents. This occurred because the search operated at the nested field level or chunk level rather than the document level.  In the worst scenario, its possible that k neighbors could be the chunks belonging to the single document, resulting in just one document in search results than desired. While increasing the k value could retrieve more documents, it introduced unnecessary search overhead.
+
+Consider an example where there are three documents labeled 1, 2, and 3, each containing two vectors in the nested KNN field. Document 1 comprises vectors A and B, Document 2 includes vectors C and F, and Document 3 contains vectors D and E. When searching for the nested KNN field, if the two nearest vectors are A and B, both belonging to Document 1, it returns only Document 1, even if the k value is set to 2.
+
+<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-1.png" alt="multi-vector-behavior-before" class="img-centered"/>
+
+### Low Recall
+The issue of fewer results mentioned earlier can also lead to low recall. In OpenSearch, an index consists of multiple shards, each containing multiple segments. Searches operate at the segment level, with results aggregated at the shard level. The results from all shards are aggregated and returned to the user finally. Consequently, if a single segment contains all of the top k documents but returns fewer than k documents, the final results, although containing k documents, would not represent the true top k documents. This is because the search algorithm operates at the segment level, potentially missing relevant documents from other segments that should have been included in the top k results.
```

Comment:
```suggestion
Returning fewer results mentioned earlier can also lead to low recall. In OpenSearch, an index consists of multiple shards, each containing multiple segments. Search operates at the segment level, with results aggregated at the shard level. After this, the results from all shards are aggregated and returned to the user. Consequently, if a single segment contains all of the top k documents but returns fewer than k documents, the final results will contain k documents but will not represent the true top k documents. This is because the search algorithm operates at the segment level, potentially missing relevant documents from other segments that should have been included in the top k results.
```